crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2189,13 +2184,25 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 
     fn infer_raise_statement(&mut self, raise: &ast::StmtRaise) {
-        let ast::StmtRaise {
-            range: _,
-            exc,
-            cause,
-        } = raise;
-        self.infer_optional_expression(exc.as_deref());
-        self.infer_optional_expression(cause.as_deref());
+        let Some(raised) = &raise.exc else {
+            return;
+        };
```

Comment:
As you've seen before in your PRs, red-knot panics if we fail to infer the types for all sub-nodes in an AST. The way you've written this method means that if `exc` is `None` but `cause` is `Some()`, red-knot will not infer the type of the `cause` sub-node, and might panic. Now, with valid Python syntax, it's impossible for `cause` to be `Some()` if `exc` is `None` -- but red-knot needs to be able to analyze invalid Python syntax without panicking as well as valid Python syntax. I think currently our parser will never produce an `ast.Raise` node where `exc` is `None` but `cause` is `Some()`, even when analyzing invalid syntax -- but that's an implementation detail of our parser as it is today; we might change the way we do error recovery at any point, so we shouldn't rely on it.

TL;DR: we can't do an early return here; you have to make sure that we also infer the type of `cause` here if it's `Some()`.